spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeConditionsInRowLevelOperations.scala

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.expressions.{Expression, Literal, SubqueryExpression}
+import org.apache.spark.sql.catalyst.plans.logical.{DeleteFromTable, Filter, LocalRelation, LogicalPlan}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2ScanRelation
+
+// we have to optimize expressions used in delete/update before we can rewrite row-level operations
+// otherwise, we will have to deal with redundant casts and will not detect noop deletes
+// it is a temp solution since we cannot inject rewrite of row-level ops after operator optimizations
+object OptimizeConditionsInRowLevelOperations extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case d @ DeleteFromTable(table, cond) if !SubqueryExpression.hasSubquery(cond.getOrElse(Literal.TrueLiteral)) =>
+      val optimizedCond = optimizeCondition(cond.getOrElse(Literal.TrueLiteral), table)
+      d.copy(condition = Some(optimizedCond))
+  }
+
+  private def optimizeCondition(condition: Expression, targetTable: LogicalPlan): Expression = {
+    val optimizer = SparkSession.active.sessionState.optimizer
+    optimizer.execute(Filter(condition, targetTable)) match {
+      case Filter(optimizedCondition, _) => optimizedCondition
```

Comment:
The first and last case make sense here to me, If we get a Filter out of our optimization we want the new condition. If we end up with DataSourcecv2ScanRelation this is essentially a "delete everything" request and we pass back true.

But if we end up with a LocalRelation why do we want to delete nothing? Is this basically saying that we ended up trying to delete but the condition applied in the delete just results in not actually effecting a DSV2 table?

Reply:
I think the `LocalRelation` would happen if 0 records matched. Spark would replace the scan + filter with an empty `LocalRelation`. That means that 0 records match the delete filter, so we can replace the whole delete filter with `false` to match 0 rows.

Similarly, if the filter is removed then it must have matched everything, so the filter should be `true`.

Reply:
ah so the opposite I thought. Ok Sounds good.

Reply:
@rdblue is correct.